rllib/core/models/base.py

```diff
@@ -0,0 +1,239 @@
+
+
+# Top level keys that unify model i/o.
+STATE_IN: str = "state_in"
+STATE_OUT: str = "state_out"
+ENCODER_OUT: str = "encoder_out"
+# For Actor-Critic algorithms, these signify data related to the actor and critic
+ACTOR: str = "actor"
+CRITIC: str = "critic"
+
+
+@ExperimentalAPI
+@dataclass
+class ModelConfig(abc.ABC):
+    """Base class for model configurations.
+
+    ModelConfigs are framework-agnostic.
+    A ModelConfig is usually built by RLModules by querying it form a Catalog object.
+    It is therefore a means of configuration for RLModules.
+    However, that is not a limitation, and they can be built directly as well.
+
+    Attributes:
+        output_dim: The output dimension of the network.
+
+    Usage Example together with a Model:
+
+    .. testcode::
+
+        class MyModel(Model):
+            def __init__(self, config):
+                super().__init__(config)
+                self.my_param = config.my_param * 2
+
+            def _forward(self, input_dict):
+                return input_dict["obs"] * self.my_param
+
+
+        @dataclass
+        class MyModelConfig(ModelConfig):
+            my_param: int = 42
+
+            def build(self, framework: str):
+                if framework == "bork":
+                    return MyModel(self)
+
+
+        config = MyModelConfig(my_param=3)
+        model = config.build(framework="bork")
+        print(model._forward({"obs": 1}))
+
+    .. testoutput::
+
+        6
+
+    """
+
+    output_dim: int = None
+
+    @abc.abstractmethod
+    def build(self, framework: str):
+        """Builds the model.
+
+        Args:
+            framework: The framework to use for building the model.
+        """
+        raise NotImplementedError
+
+
+class Model(abc.ABC):
+    """Framework-agnostic base class for RLlib models.
+
+    Models are low-level neural network components that offer input- and
+    output-specification, a forward method, and a get_initial_state method. Models
+    are composed in RLModules.
+
+    Usage Example together with ModelConfig:
+
+    .. testcode::
+
+        class MyModel(Model):
+            def __init__(self, config):
+                super().__init__(config)
+                self.my_param = config.my_param * 2
+
+            def _forward(self, input_dict):
+                return input_dict["obs"] * self.my_param
+
+
+        @dataclass
+        class MyModelConfig(ModelConfig):
+            my_param: int = 42
+
+            def build(self, framework: str):
+                if framework == "bork":
+                    return MyModel(self)
+
+
+        config = MyModelConfig(my_param=3)
+        model = config.build(framework="bork")
+        print(model._forward({"obs": 1}))
+
+    .. testoutput::
+
+        6
+
+    """
+
+    def __init__(self, config: ModelConfig):
+        self.config = config
+        # Model itself does not impose restrictions on inputs and outputs.
+        self.input_spec = DummySpec()
```

Comment:
This is in the last file of this diff.
Please let me know if you can think of a more elegant solution.
Requirements: Make sure that a model without any requirements to inputs/outputs does not error out when we check specs on the forward method.

Reply:
The thing is that SpecDict barfs if it does not get a mapping to check.
My first reaction was that always requiring a mapping as output for all models is not a restriction we want to impose on people.
Especially folks who just want to write some custom models would be weirded out by the spec checking errors this produces if they specifically have no specs defined.

Reply:
how does convert_to_canonical work when an empty list is given to it? We want the output of that, right?

Reply:
here: https://sourcegraph.com/github.com/ray-project/ray/-/blob/rllib/models/specs/checker.py?L82&subtree=true
the output would be `SpecDict({})`

Reply:
Huh. I was not aware of that tool (`convert_to_canonical_format`), sorry.
Let me read through that first and see if I can apply it here.

Reply:
```
from ray.rllib.core.models.base import Model, ModelConfig
from ray.rllib.models.specs.checker import check_input_specs
from ray.rllib.models.specs.checker import check_output_specs
from dataclasses import dataclass

class MyModel(Model):
    def __init__(self, config):
        super().__init__(config)
        self.my_param = config.my_param * 2

    @check_input_specs("input_spec")
    @check_output_specs("output_spec")
    def _forward(self, input_dict):
        return input_dict["obs"] * self.my_param


@dataclass
class MyModelConfig(ModelConfig):
    my_param: int = 42

    def build(self, framework: str):
        if framework == "bork":
            return MyModel(self)


config = MyModelConfig(my_param=3)
model = config.build(framework="bork")
print(model._forward({"obs": 1}))
```

This code results in the following error:

```
File "/Users/artur/code/ray/python/ray/rllib/models/specs/checker.py", line 152, in _validate
    raise ValueError(f"{tag} must be a Mapping, got {type(data).__name__}")
ValueError: output must be a Mapping, got int
```

Reply:
.... with:
```
Model.input_spec = convert_to_canonical_format([])
Model.output_spec = convert_to_canonical_format([])
```

Reply:
looking ...

Reply:
After offline discussion, both checker decorators need to only check specs if the specs are not None. So if we make the base class by default return None, then the specs won't get checked and user won't have to worry about it.

Reply:
This will get rid of needing a DummySpec()